src/D2L.Security.OAuth2/Keys/Default/RotatingPrivateKeyProvider.cs

```diff
@@ -25,6 +26,7 @@ TimeSpan keyRotationPeriod
 			m_keyRotationPeriod = keyRotationPeriod;
 		}
 
+		[GenerateSync]
 		async Task<D2LSecurityToken> IPrivateKeyProvider.GetSigningCredentialsAsync() {

```

Comment:
This function loses a critical `try` because it only has a `finally` block

```csharp
		[Blocking]
D2LSecurityToken IPrivateKeyProvider.GetSigningCredentials() {

			// Hold a local reference so that we know we are talking about the same key
			// after even if another thread changed m_privateKey (race condition when we
			// are using a key very close to the rotation time.)
			D2LSecurityToken privateKey = m_privateKey;

			if( NeedFreshPrivateKey( privateKey ) ) {
				// This Semaphore is used instead of lock(foo){}
				// because await cannot be used within a lock
				m_privateKeyLock.Wait();
{
					privateKey = m_privateKey;

					if( NeedFreshPrivateKey( privateKey ) ) {
						m_privateKey = ( m_inner.GetSigningCredentials()).Ref();

						if( privateKey != null ) {
							privateKey.Dispose();
						}

						privateKey = m_privateKey;
					}

				} 			}

			return privateKey.Ref();
		}
```

Reply:
Oh, very good catch! I will fix that in the generator

Reply:
Spicy! Nice catch.